### PR TITLE
Feature/DHH-830

### PR DESCRIPTION
**Pull request by hweng**

### CODE REVIEW

**Comment by hweng** [comment]
@mcritchlow  The new added implementation would load and cache the share config on the first request, and for the subsequent requests it would just fetch the value from memory.

The share_config.yml is updated to use typical rails environment configuration file.

The #replace_config_file, #copy_config_file, #write_file methods and share_notify.yml.prod.sample, share_notify.yml.staging.sample files from new_deploy_solution pull request are all removed.

**Comment by mcritchlow** [comment]
@hweng - thanks, will review now (and watch for CirceCI status)

@ucsdlib/developers - please review, comment, etc.

File: app/controllers/dams_resource_controller.rb [hunk 1]
```diff
@@ -229,14 +213,24 @@ def osf_push
...
     if document.valid?
-      replace_config_file ('share_notify.yml')
-      api = ShareNotify::API.new
-      api.post(document.to_share.to_json)
-      copy_config_file ('share_notify.yml') if !Rails.env.production?
-      redirect_to dams_collection_path(params[:id]), notice: "Your record has been pushed to OSF Share Staging area."
+      @headers = {'Authorization' => "Token #{share_config.fetch('token', nil)}", 
+                 'Content-Type'  => 'application/json'
+                }
+      @route = "#{share_config.fetch('host', 'https://staging.osf.io/')}api/v1/share/data"
```
Comment (mcritchlow) [inline app/controllers/dams_resource_controller.rb]: Do we need a default here from the `#fetch` call? It looks like the config is already providing defaults.

File: config/share_notify.yml [hunk 2]
```diff
@@ -1,3 +1,30 @@
...
+  <<: *defaults
+
+demo:
+  <<: *defaults
+
+test:
+  <<: *defaults
+
+production:
+  host: <%= ENV.fetch('APPS_DHH_PROD_SHARE_NOTIFY_HOST') {'https://staging.osf.io/'}%>
```
Comment (mcritchlow) [inline config/share_notify.yml]: do we want production to fallback to staging? this seems potentially problematic. I wonder, in the scenario where production doesn't have the proper environment variables, if it wouldn't be better to have it fail with a KeyError so we know something is wrong?

**Comment by mcritchlow** [comment]
overall, 👍 from me on this barring a couple questions on the configuration. nice!

**Comment by hweng** [comment]
@mcritchlow I removed the default values from share_config. It doesn't cause any problem or hit performance if we leave it there though. I think one default value in the share method is good enough.

**Comment by VivianChu** [comment]
@hweng  The changes look great.  I checked out your branch and got 1 error 

Failures:

  1) Vistor wants to push a record to OSF Share Staging area should call ShareNotify
     Failure/Error: @response = with_timeout { HTTParty.post(@route, body: document.to_share.to_json, headers: @headers)}
     
     SocketError:
       getaddrinfo: nodename nor servname provided, or not known
     # ./app/controllers/dams_resource_controller.rb:221:in `block in osf_push'
     # ./app/controllers/dams_resource_controller.rb:235:in `block in with_timeout'

If nobody else and circleci don't have that error, then it's fine with me.

**Comment by hweng** [comment]
@VivianChu I don't have that error on my local. And circleci don't have that error too.  Let's see if @mcritchlow or @lsitu get the error from their local.  Would you pull the latest version of the branch and rerun the test? And if you still got the error, it could be your local environment issue which I can help you figure out.

**Comment by VivianChu** [comment]
@hweng  That's the latest version.  I did the pull only 30 minutes ago when your last commit was
 last night. This is my first time seeing that error. If @lsitu & @mcritchlow  don't get that error then 👍 for this PR.

**Comment by lsitu** [comment]
@VivianChu: @hweng I don't see that error but got another weird error that may not be related: 
Failures:

  1) complex object view testing component pager functionality
     Failure/Error: visit dams_object_path(@damsComplexObj.pid)
     
     Capybara::Poltergeist::StatusFailError:
       Request to 'http://127.0.0.1:54087/object/xx97626129' failed to reach server, check DNS and/or server status
     # ./spec/features/dams_object_spec.rb:552:in `block (2 levels) in <top (required)>'

**Comment by hweng** [comment]
@lsitu Thank you for testing, which it is passed for you. Yes, that one is not related to this pull request.

**Comment by hweng** [comment]
@VivianChu I think the problem in your local environment is that you might store the incorrect host value in you .bashrc. To solve your local issue, you can either remove the environment variable in your .bashrc which the config would fall to default, or update your .bashrc to use correct value.

**Comment by VivianChu** [comment]
@hweng I removed those environment variables from my .bashrc and it works.

**Comment by hweng** [comment]
@VivianChu Great! It works for you.

**Comment by VivianChu** [comment]
@hweng The value for this variable  APPS_DHH_STAGING_NOTIFY_TOKEN in my .bashrc was "default" before, not "share_token".  That's why it was failing.

**Comment by hweng** [comment]
@VivianChu No, It doesn't matter whether the token value is "default" or "share_token", which are both fake value. But the host name does matter which is the reason that you got failed on your local.
I'm glad that it works fine on your local now.

**Comment by VivianChu** [comment]
@hweng  The value for the host name was https://staging.osf.io in my local .bashrc.  Anyway, I'm glad that it works now.

**Comment by hweng** [comment]
@VivianChu The correct value is "https://staging.osf.io/", or no env vars in your local .bashrc, both would work.

**Comment by VivianChu** [comment]
@hweng I'm very sure that I got the correct value for the host.  I just restored my .bashrc with that old host and the token 'default' and the tests still passed. I think it's just intermittent error that can't be replicate all the time. So whenever it ever happened again, I will just ignore it. 👍 for this PR

**Comment by hweng** [comment]
@VivianChu So after you restore your .bashrc, did you restart your terminal?  You know it is very hard to help debugging your local environment here.
Since the tests passed on CirceCI, and passed for all of our local. Can anyone merge the pull request?

**Comment by VivianChu** [comment]
@hweng   Yes I did restart my terminal.

**Comment by hweng** [comment]
@VivianChu Then I don't know. I can not replicate your local environment.  Thanks for merging!

**Pull request merged**